File: src/ffi/alerts/file_error.rs
#[cxx::bridge(namespace = "ltrs")]
pub(crate) mod ffi {
    #[namespace = "libtorrent"]
    extern "C++" {
        type file_error_alert = crate::ffi::ffi::file_error_alert;
    }

    unsafe extern "C++" {
        include!("cpp/alerts/file_error.h");
        type Error = crate::ffi::error::ffi::Error;

        unsafe fn file_error_alert_get_filename<'a>(alert: *mut file_error_alert) -> &'a str;
        unsafe fn file_error_alert_get_error(alert: *mut file_error_alert) -> Error;
        unsafe fn file_error_alert_get_op(alert: *mut file_error_alert) -> u8;
    }
}

File: cpp/alerts/file_error.cpp
#include "./file_error.h"

namespace ltrs {

rust::Str file_error_alert_get_filename(lt::file_error_alert* alert) {
    return alert->filename();
}

Error file_error_alert_get_error(lt::file_error_alert* alert) {
    return error_code_to_error(alert->error);
}

uint8_t file_error_alert_get_op(lt::file_error_alert* alert) {
    return static_cast<uint8_t>(alert->op);
}

} // namespace ltrs
